synapse/storage/databases/main/roommember.py

```diff
@@ -384,14 +388,19 @@ async def get_rooms_for_local_user_where_membership_is(
             self._get_rooms_for_local_user_where_membership_is_txn,
             user_id,
             membership_list,
+            excluded_rooms,
         )
 
         # Now we filter out forgotten rooms
         forgotten_rooms = await self.get_forgotten_rooms_for_user(user_id)
         return [room for room in rooms if room.room_id not in forgotten_rooms]
```

Comment:
Out of curiosity, is there any reason that the excluded rooms couldn't be filtered out here, like the forgotten rooms are, rather than changing the query to the database in `_get_rooms_for_local_user_where_membership_is_txn`?

Reply:
My rationale here is that we already know from the start which rooms we want to exclude, so ignoring them from the result would minimise the work and keep the memory usage to a minimum. I'm not sure why we're not already doing that for forgotten rooms, I'll ask around.

Reply:
According to #synapse-dev, it's likely a combination of `NOT IN` not being very well supported at the time and the forgotten rooms list being user-defined and not capped thus it's possible it becomes large enough to cause issues if used directly in the query. I think we're good on both points with the excluded rooms, since it's an admin-defined static list that should remain fairly small, but I'll still change it just to simplify the query and the code (and it shouldn't make a huge difference).